src/Network/NWFramerMessage.cs

```diff
@@ -0,0 +1,106 @@
+//
+// NWFramer.cs: Bindings the Network nw_framer_t API.
+//
+// Authors:
+//   Manuel de la Pena (mandel@microsoft.com)
+//
+// Copyright 2019 Microsoft Inc
+//
+using System;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+using ObjCRuntime;
+using Foundation;
+using CoreFoundation;
+
+using OS_nw_framer=System.IntPtr;
+using OS_nw_protocol_metadata=System.IntPtr;
+using OS_dispatch_data=System.IntPtr;
+using OS_nw_protocol_definition=System.IntPtr;
+using OS_nw_protocol_options=System.IntPtr;
+using OS_nw_endpoint=System.IntPtr;
+using OS_nw_parameters=System.IntPtr;
+
+namespace Network {
+
+	[TV (13,0), Mac (10,15), iOS (13,0), Watch (6,0)]
+	public class NWFramerMessage : NWProtocolMetadata {
+		internal NWFramerMessage (IntPtr handle, bool owns) : base (handle, owns) {}
+
+		[DllImport (Constants.NetworkLibrary)]
+		static extern unsafe void nw_framer_message_set_value (OS_nw_protocol_metadata message, string key, byte *value, void *dispose_value);
+
+		public void SetData (string key, ReadOnlySpan<byte> value)
+		{
+			// the method takes a callback to cleanup the data, but we do not need that since we are managed
+			if (key == null)
+				throw new ArgumentNullException (nameof (key));
+
+			unsafe {
+				fixed (byte* mh = value)
+					nw_framer_message_set_value (GetCheckedHandle (), key,  mh, null);
```

Comment:
This is not safe: you're storing a pointer to memory referenced by the ReadOnlySpan, but the ReadOnlySpan will be collected soon after this function returns, and that might include the memory pointed to by the ReadOnlySpan.

You must keep the pointer passed to the framer message alive for as long needed (until either the value is overwritten or the framer message freed).

As such I don't think it's a good idea to use a ReadOnlySpan as the parameter type, because it can only live on the stack and there's no way for this code to make sure it's not collected once the call stack unwinds. I think an IntPtr version could work, but that would also require the API to have a dispose callback to notify the caller when the IntPtr can be freed:

```csharp
public void SetData (string key, IntPtr value, Action<IntPtr> dispose_value);
```

another alternative is a byte[] version:

```csharp
public void SetData (string key, byte[] data);
```

and where you store a strong GCHandle to the byte[] in the framer message (and free the GCHandle in the dispose value callback).

Reply:
byte[] using the cleanup in the c call is doable and probably the way to go.